(#343) Compound parent bugs & parent(GameObject g, boolean compound)

rafcolson opened the pull request:
**Fix: Compound Parent(null) crash -> remove/add the parent body**

- Bug: crash when unparenting a compound parent.

**Fix: excessive Compound margin -> set to zero**

- Bug: The total collision margin of a compound object (and its children) turns out to be higher as the margin in the Blender Collision tab. It's the total of two margins. The first is the margin of the compound object itself, the second is the margin of the children. If not set, the margin value will be 0.04, so it's necessary to set the margin of the compound object to zero.

**Fix: Compound child collision detection -> remove body**

- When a collision object is parented to a compound object, the shape of that object is added to the compound shape of the parent, leaving it obsolete. This causes collision interference when the child is not ghost and redundant physics calculation. Solution is to remove their bodies from the world when compound parenting.
- At initialization the children are hooked to their parents before they are added to the world for the first time, initializing collision groups and masks also. A private addToWorldNoChildren() enables hookParentChild() to happen after this. This changes nothing else. Initialization of collision groups and masks would get complicated if not doing this while adding to the world, so this seems to be the best solution.

**Fix: replaceModel broken with Compound child -> unparent**

- Bug: replaceModel didn't support physics update if compound parented.

GoranM commented:
> Initialization of collision groups and masks would get complicated if not doing this while adding to the world, so this seems to be the best solution.

In what way? Please describe the actual problem that your changes are intended to solve.

Also, please remember that commit messages should be concise. Try to condense things down to a single sentence: What you did, and why.

Don't write paragraphs - If you really need to elaborate on something, you can do it in the PR.

rafcolson commented:
>In what way? Please describe the actual problem that your changes are intended to solve.

As far as I know it's not possible to set the group or mask unless objects are added to the world already. Unless I'm wrong and it is possible, but that would be making things quite complicated, by manually constructing the BroadfaseProxy's. I've looked into it a couple of days, but couldn't get it to work.

But all of this is aside of what's the actual intention of the commit to solve the problem, which is: *At initialization the children are hooked to their parents before they are added to the world for the first time, initializing collision groups and masks also. A private addToWorldNoChildren() enables hookParentChild() to happen after this.* In other words, in order to have compound parenting work at initialization, this has to happen *after* all instances are added to the scene. Else, objects with collision (and not ghost) would be in the world, while they should remain out of the world.

rafcolson commented:
>Also, please remember that commit messages should be concise. Try to condense things down to a single sentence: What you did, and why.

I'll try to find more descriptive titles for those commits. Those paragraphs where intended to describe the commit content in detail. But if it's not really helpful, I can leave those out the next time.

rafcolson commented:
PR updated.

rafcolson commented:
I would like to put this on hold until #347 has been discussed.

rafcolson commented:
PR updated. Note that I removed **Fix: NO_COLLISION or SENSOR Compound child -> only if not** and added method ```GameObject.parent(GameObject p, boolean compound)``` to open the possibility to disable compound parenting if wanted. Not really a fix, but anyhow, I think I can add it to the PR.

GoranM commented:
> Do you know how I could change those titles directly? I'm always doing a git reset HEAD^ to get to the previous commits and do git commit --amend. This requires me to save every commit in a text file in order to restore. If you know a better way, that would be really helpful.

https://help.github.com/articles/changing-a-commit-message/

rafcolson commented:
Thanks, that's very useful. I must have looked over this before.
